pkg/reconciler/buildrun/runner/taskrun.go

```diff
@@ -0,0 +1,491 @@
+
+		for _, result := range taskRunResults {
+			if result.Key == prefixedResultErrorMessage {
+				errorMessage = result.Value
+			}
+
+			if result.Key == prefixedResultErrorReason {
+				errorReason = result.Value
+			}
+		}
+	}
+
+	return errorReason, errorMessage
+}
+
+func extractFailedPodAndContainer(ctx context.Context, client client.Client, taskRun *pipelineapi.TaskRun) (*corev1.Pod, *corev1.Container, *corev1.ContainerStatus, error) {
+	var pod corev1.Pod
+	if err := client.Get(ctx, types.NamespacedName{Namespace: taskRun.Namespace, Name: taskRun.Status.PodName}, &pod); err != nil {
+		return nil, nil, nil, err
+	}
+
+	failures := make(map[string]*corev1.ContainerStatus)
+	// Find the names of all containers with failure status
+	for i := range pod.Status.ContainerStatuses {
+		containerStatus := pod.Status.ContainerStatuses[i]
+		if containerStatus.State.Terminated != nil && containerStatus.State.Terminated.ExitCode != 0 {
+			failures[containerStatus.Name] = &containerStatus
+		}
+	}
+
+	// Find the first container that has a failure status
+	var failedContainer *corev1.Container
+	var failedContainerStatus *corev1.ContainerStatus
+	for i, container := range pod.Spec.Containers {
+		if containerStatus, has := failures[container.Name]; has {
+			failedContainer = &pod.Spec.Containers[i]
+			failedContainerStatus = containerStatus
+			break
+		}
+	}
+
+	return &pod, failedContainer, failedContainerStatus, nil
+}
+
+// updateBuildRunUsingTaskFailures is extracting failures from taskRun steps and adding them to buildRun (mutates)
+func (t *TaskRunBuildRunner) updateBuildRunUsingTaskFailures(ctx context.Context, client client.Client, buildRun *v1beta1.BuildRun) {
+	trCondition := t.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
+
+	// only extract failures when failing condition is present
+	if trCondition != nil && pipelineapi.TaskRunReason(trCondition.Reason) == pipelineapi.TaskRunReasonFailed {
+		buildRun.Status.FailureDetails = extractFailureDetails(ctx, client, t.TaskRun)
+	}
+}
+
+type TaskRunBuildRunnerFactory struct {
+	config *config.Config
+	scheme *runtime.Scheme
+}
+
+// NewTaskRunBuildRunnerFactory creates a new TaskRunBuildRunnerFactory instance.
+func NewTaskRunBuildRunnerFactory(scheme *runtime.Scheme, controllerConfig *config.Config) BuildRunnerFactory {
+	return &TaskRunBuildRunnerFactory{
+		scheme: scheme,
+		config: controllerConfig,
+	}
+}
+
+func (t *TaskRunBuildRunnerFactory) CreateBuildRunner(serviceAccount *corev1.ServiceAccount, strategy v1beta1.BuilderStrategy, build *v1beta1.Build, buildRun *v1beta1.BuildRun) (BuildRunner, error) {
+	var (
+		generatedTaskRun *pipelineapi.TaskRun
+	)
+
+	generatedTaskRun, err := resources.GenerateTaskRun(t.config, build, buildRun, serviceAccount.Name, strategy)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Set OwnerReference for BuildRun and TaskRun
+	if err := controllerutil.SetOwnerReference(buildRun, generatedTaskRun, t.scheme); err != nil {
+		return nil, err
+	}
```

Comment:
Leaving this here for now, but I may move the setting of owner references to the main Reconcile function in a subsequent PR.